shared/subscribers/service_test.go

```diff
@@ -0,0 +1,148 @@
+package subscribers
+
+import (
+	"context"
+	"testing"
+
+	"code.vegaprotocol.io/vega/core/events"
+	dtypes "code.vegaprotocol.io/vega/core/types"
+	"code.vegaprotocol.io/vega/logging"
+	eventspb "code.vegaprotocol.io/vega/protos/vega/events/v1"
+	"code.vegaprotocol.io/vega/shared/subscribers/mocks"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSlowConsumerIsDisconnected(t *testing.T) {
+	ctx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+
+	ctrl := gomock.NewController(t)
+
+	broker := mocks.NewMockBroker(ctrl)
+	broker.EXPECT().Subscribe(gomock.Any())
+	broker.EXPECT().Unsubscribe(gomock.Any())
+
+	maxBufferSize := 3
+	testStreamSubscription := TestStreamSubscription{events: make(chan []*eventspb.BusEvent)}
+
+	var serviceContext context.Context
+	s := NewService(logging.NewTestLogger(), broker, maxBufferSize)
+	out, _ := s.ObserveEventsOnStream(ctx, 2, func(passedCtx context.Context) StreamSubscription {
+		serviceContext = passedCtx
+		return testStreamSubscription
+	})
+
+	testStreamSubscription.events <- []*eventspb.BusEvent{
+		events.NewAccountEvent(ctx, dtypes.Account{
+			ID: "acc-1",
+		}).StreamMessage(),
+		events.NewAccountEvent(ctx, dtypes.Account{
+			ID: "acc-2",
+		}).StreamMessage(),
+	}
+
+	events1 := <-out
+	assert.Equal(t,
+		[]*eventspb.BusEvent{
+			events.NewAccountEvent(ctx, dtypes.Account{ID: "acc-1"}).StreamMessage(),
+			events.NewAccountEvent(ctx, dtypes.Account{ID: "acc-2"}).StreamMessage(),
+		},
+		events1)
+
+	testStreamSubscription.events <- []*eventspb.BusEvent{
+		events.NewAccountEvent(ctx, dtypes.Account{
+			ID: "acc-3",
+		}).StreamMessage(),
+		events.NewAccountEvent(ctx, dtypes.Account{
+			ID: "acc-4",
+		}).StreamMessage(),
+		events.NewAccountEvent(ctx, dtypes.Account{
+			ID: "acc-5",
+		}).StreamMessage(),
+		events.NewAccountEvent(ctx, dtypes.Account{
+			ID: "acc-6",
+		}).StreamMessage(),
+	}
+
+loop:
+	for {
+		select {
+		case <-serviceContext.Done():
+			break loop
+		}
+	}
+
+	events1, _ = <-out
+	assert.Equal(t,
+		[]*eventspb.BusEvent{
+			events.NewAccountEvent(ctx, dtypes.Account{ID: "acc-3"}).StreamMessage(),
+			events.NewAccountEvent(ctx, dtypes.Account{ID: "acc-4"}).StreamMessage(),
+			events.NewAccountEvent(ctx, dtypes.Account{ID: "acc-5"}).StreamMessage(),
+			events.NewAccountEvent(ctx, dtypes.Account{ID: "acc-6"}).StreamMessage(),
+		},
+		events1)
+
+	// We expect this channel to close
+	for range out {
```

Comment:
If the channel doesn't get closed, this is going to hang.

Reply:
true, it should always close, if not we have bug and we'll know about it cause the test times out, does it need more?  it should never happen

Reply:
The timeout is 30 whole minutes, and then when it does timeout it won't tell you why

Anyway I think the test could be better if we `EXPECT()` on the connection being unsubscribed. Its also a bit confusing and not accurate that the subscriber _is_ reading quick enough but we just artificially send too much into the channel. Something like this would be better:
```
// send in 4 things, one sitting waiting in the out channel and 3 in the data so we're at buffer max
	for i := 0; i < 4; i++ {
		testStreamSubscription.events <- []*eventspb.BusEvent{
			events.NewAccountEvent(ctx, dtypes.Account{
				ID: "acc-1",
			}).StreamMessage(),
		}
	}

	// this one will fill up the buffer too much causes unsubscription from the broker
	broker.EXPECT().Unsubscribe(gomock.Any()).Times(1)
	testStreamSubscription.events <- []*eventspb.BusEvent{
		events.NewAccountEvent(ctx, dtypes.Account{
			ID: "acc-3",
		}).StreamMessage(),
	}
	<-serviceContext.Done()
```

Reply:
btw you do get the stack on timeout:


goroutine 18 [sleep]:
time.Sleep(0x22ecb25c000)
	/usr/local/go/src/runtime/time.go:195 +0x135
code.vegaprotocol.io/vega/shared/subscribers.TestSlowConsumerIsDisconnected(0xc000291860?)
	/home/matthewpendrey/projects/vegadevelop/vega/shared/subscribers/service_test.go:66 +0xf17
testing.tRunner(0xc000291860, 0xde4310)
	/usr/local/go/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1493 +0x35f
FAIL	code.vegaprotocol.io/vega/shared/subscribers	60.070s
?   	code.vegaprotocol.io/vega/shared/subscribers/mocks	[no test files]
FAIL

Reply:
Then do this:
```
wait, cancel := context.WithCancel(context.Background())
	broker.EXPECT().Unsubscribe(gomock.Any()).Times(1).Do(
		func(_ interface{}) {
			cancel()
		},
	)
	testStreamSubscription.events <- []*eventspb.BusEvent{
		events.NewAccountEvent(ctx, dtypes.Account{
			ID: "acc-3",
		}).StreamMessage(),
	}

	select {
	case <-wait.Done():
	case <-time.After(2 * time.Second):
		assert.Fail(t, "broker never unsubscribed")
	}
```

Reply:
think that's at risk of intermittent timing failure under race test isn't ?  the original seems more robust

Reply:
How?